### PR TITLE
Make `expected<any, ...>` and `expected<void, any>` copyable

### LINKED ISSUE #4011: `<expected>`: `std::function` returning `expected<any, int>` emits error C7608: atomic constraint should be a constant expression
Originally reported by Jared Resch as https://godbolt.org/z/YTe7hYPrq . I've further reduced this to https://godbolt.org/z/Ebadxz43K which is accepted by libstdc++ and libc++, but rejected by VS 2022 17.8 Preview 1:

```
C:\Temp>type meow.cpp
```
```cpp
#include <any>
#include <expected>
#include <functional>
using namespace std;

using X = expected<any, int>;

X meow() {
    return {};
}

int main() {
    function<X()> f{meow};
}
```
```
C:\Temp>cl /EHsc /nologo /W4 /std:c++latest meow.cpp
meow.cpp
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\expected(258): error C7608: atomic constraint should be a constant expression
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\expected(258): note: the template instantiation context (the oldest one first) is
meow.cpp(13): note: while processing the default template argument of 'std::function<X (void)>::function(_Fx &&)'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\functional(1050): note: see reference to alias template instantiation 'std::_Func_class<_Ret>::_Enable_if_callable_t<X(__cdecl &)(void),std::function<X (void)>>' being compiled
        with
        [
            _Ret=X
        ]
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\functional(892): note: see reference to variable template 'const bool conjunction_v<std::negation<std::is_same<std::expected<std::any,int> __cdecl(void),std::function<std::expected<std::any,int> __cdecl(void)> > >,std::_Is_invocable_r<std::expected<std::any,int>,std::expected<std::any,int> (__cdecl*&)(void)> >' being compiled
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\functional(892): note: see reference to class template instantiation 'std::_Is_invocable_r<_Ret,X (__cdecl *&)(void)>' being compiled
        with
        [
            _Ret=X
        ]
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\type_traits(1889): note: see reference to alias template instantiation 'std::_Is_invocable_r_<_Rx,_Callable,>' being compiled
        with
        [
            _Rx=X,
            _Callable=X (__cdecl *&)(void)
        ]
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\type_traits(1886): note: see reference to alias template instantiation 'std::_Invoke_traits_common<std::expected<std::any,int>,false>::_Is_invocable_r<_Rx>' being compiled
        with
        [
            _Rx=X
        ]
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\type_traits(1810): note: see reference to variable template 'const bool disjunction_v<std::is_void<std::expected<std::any,int> >,std::_Invoke_convertible<std::expected<std::any,int>,std::expected<std::any,int>,void> >' being compiled
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\expected(271): note: see reference to variable template 'const bool std::expected<std::any,int>::_Allow_unwrapping<std::any,int>' being compiled
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\expected(242): note: see reference to variable template 'const bool disjunction_v<std::is_same<std::any,bool>,std::negation<std::disjunction<std::is_constructible<std::any,std::expected<std::any,int> &>,std::is_constructible<std::any,std::expected<std::any,int> >,std::is_constructible<std::any,std::expected<std::any,int> const &>,std::is_constructible<std::any,std::expected<std::any,int> const >,std::is_convertible<std::expected<std::any,int> &,std::any>,std::is_convertible<std::expected<std::any,int> &&,std::any>,std::is_convertible<std::expected<std::any,int> const &,std::any>,std::is_convertible<std::expected<std::any,int> const &&,std::any> > > >' being compiled
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\expected(250): note: see reference to class template instantiation 'std::is_constructible<_Ty,std::expected<_Ty,int> &>' being compiled
        with
        [
            _Ty=std::any
        ]
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\type_traits(755): note: while processing the default template argument of 'std::any::any(_ValueType &&)'
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\any(138): note: see reference to variable template 'const bool conjunction_v<std::negation<std::is_same<std::expected<std::any,int>,std::any> >,std::negation<std::_Is_specialization<std::expected<std::any,int>,std::in_place_type_t> >,std::is_copy_constructible<std::expected<std::any,int> > >' being compiled
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\expected(258): error C2131: expression did not evaluate to a constant
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\expected(258): note: failure was caused by a read of an uninitialized symbol
C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.38.32919\include\expected(258): note: see usage of 'std::expected<std::any,int>::_Allow_unwrapping<std::any,int>'
```

I haven't figured out the root cause yet (library bug? compiler bug? an LWG issue resolution we've implemented that the others haven't?).

### PR DESCRIPTION
**Pull request by achabense**
Fixes #4011.

### CODE REVIEW

**Comment by achabense** [comment]
For extra verification
```cpp
#include<any>
#include<expected>
#include<functional>

using namespace std;

void test_A() {
    using T = expected<any, int>;
    T t1;
    T t2{ move(t1) };
    T t3{ t2 };

    using U = expected<void, any>;
    U u1;
    U u2{ move(u1) };
    U u3{ u2 };
}

using X = expected<any, int>;

X meow() {
    return {};
}

void test_B() {
    function<X()> f{ meow };
}

int main() {
    test_A();
    test_B();
}
```

File: stl/inc/expected [hunk 1]
```diff
@@ -252,9 +256,10 @@ public:
                                            && !is_constructible_v<unexpected<_Err>, const expected<_Uty, _UErr>&> //
                                            && !is_constructible_v<unexpected<_Err>, const expected<_Uty, _UErr>>;
 
+public:
     template <class _Uty, class _UErr>
-        requires is_constructible_v<_Ty, const _Uty&> && is_constructible_v<_Err, const _UErr&>
-              && _Allow_unwrapping<_Uty, _UErr>
+        requires _Not_self<_Uty, _UErr> && is_constructible_v<_Ty, const _Uty&>
```
Comment (CaseyCarter) [inline stl/inc/expected:261]: We conventionally use `_Different_from` (which is the Standard's exposition-only concept _`different-from`_) for this. Also, thanks for adding more fuel to my "every converting constructor needs to be constrained to not act like a special member function" rule of thumb ;)

File: tests/std/tests/P0323R12_expected/test.cpp [hunk 2]
```diff
@@ -2135,6 +2136,11 @@ void test_lwg_3843() {
     }
 }
 
+void test_gh_4011() {
```
Comment (CaseyCarter) [inline tests/std/tests/P0323R12_expected/test.cpp:2139]: 1. We like to have a comment describing what problem a regression test is trying not to regress for the day the bug database inevitably goes away.
2. This doesn't need to be run, or even in a function; we could just put these `static_assert`s at namespace scope.

**Review by CaseyCarter — APPROVED** [review]
Some comments that I'll test-and-push changes for.

File: stl/inc/expected [hunk 3]
```diff
@@ -190,6 +190,21 @@ class expected {
     template <class _UTy, class _UErr>
     friend class expected;
 
+    template <class _Uty, class _UErr>
+    static constexpr bool _Allow_unwrapping = disjunction_v<is_same<remove_cv_t<_Ty>, bool>,
```
Comment (StephanTLavavej) [inline stl/inc/expected:194]: We conventionally avoid relying on `class`'s default of `private` access control. Here, `class expected` began with only `static_assert`s and `friend`, neither of which actually care about access control, so as a special exception we don't bother to say `private:`. However, by moving `_Allow_unwrapping` up, an actual thing is being defined here, so we should say `private:`.

File: stl/inc/expected [hunk 4]
```diff
@@ -1166,6 +1167,12 @@ class expected<_Ty, _Err> {
     template <class _UTy, class _UErr>
     friend class expected;
 
+    template <class _Uty, class _UErr>
+    static constexpr bool _Allow_unwrapping = !is_constructible_v<unexpected<_Err>, expected<_Uty, _UErr>&>
```
Comment (StephanTLavavej) [inline stl/inc/expected:1171]: Ditto `private:`.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks! @CaseyCarter I pushed a trivial stylistic change after you approved.

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

File: stl/inc/expected [hunk 5]
```diff
@@ -238,23 +254,8 @@ public:
...
-                                                      is_constructible<_Ty, const expected<_Uty, _UErr>&>, //
-                                                      is_constructible<_Ty, const expected<_Uty, _UErr>>, //
-                                                      is_convertible<expected<_Uty, _UErr>&, _Ty>, //
-                                                      is_convertible<expected<_Uty, _UErr>&&, _Ty>, //
-                                                      is_convertible<const expected<_Uty, _UErr>&, _Ty>, //
-                                                      is_convertible<const expected<_Uty, _UErr>&&, _Ty>>>> //
-                                           && !is_constructible_v<unexpected<_Err>, expected<_Uty, _UErr>&> //
-                                           && !is_constructible_v<unexpected<_Err>, expected<_Uty, _UErr>> //
-                                           && !is_constructible_v<unexpected<_Err>, const expected<_Uty, _UErr>&> //
-                                           && !is_constructible_v<unexpected<_Err>, const expected<_Uty, _UErr>>;
```
Comment (achabense) [inline stl/inc/expected:253]: 1. I think it's beneficial not to relocate `_Allow_unwrapping` here and in `expected<void, ...>`, as it is used closely after definition so its function can be more obvious.
2. These can also be private:
https://github.com/microsoft/STL/blob/6c69a73911b33892919ec628c0ea5bbf0caf8a6a/stl/inc/expected#L351
https://github.com/microsoft/STL/blob/6c69a73911b33892919ec628c0ea5bbf0caf8a6a/stl/inc/expected#L363

**Comment by achabense** [comment]
I find another weird case... Should `t` contain `int or U` in this case?
```cpp
#include<any>
#include<expected>
using namespace std;

int main() {
    using T = expected<any, int>;
    using U = expected<int, int>;

    T t{ U{} };
    (void) any_cast<U>(t.value()); // not thrown, t.any contains U instead of int
}
```

**Comment by frederick-vs-ja** [comment]
> I find another weird case... `t` contains `U` instead of `int` after construction; I think this is a bit counterintuitive... [(gcc and clang also think `t` should contain `U`)](https://godbolt.org/z/r48336xMj)
> 
> ```c++
> #include<any>
> #include<expected>
> using namespace std;
> 
> int main() {
>     using T = expected<any, int>;
>     using U = expected<int, int>;
> 
>     T t{ U{} };
>     (void) any_cast<U>(t.value()); // not thrown, t.any contains U instead of int
> }
> ```

This is probably mandatory as `is_constructible_v<unexpected<any>, expected<int, int>>` is `true`. An LWG issue might be welcome but I currently have no idea on the resolution...

**Comment by StephanTLavavej** [comment]
Thanks for fixing this constraint recursion! :repeat: :hammer_and_wrench: :tada:

**Pull request merged**